pytential/qbx/cost.py

```diff
@@ -167,747 +184,791 @@ def taylor_translation_cost_model(dim, nlevels):
-        as well as calibration constants.
+    * To calibrate the model, pass operation counts per stage together with timing
+      data to :meth:`estimate_kernel_specific_calibration_params`.
 
-    .. automethod:: copy
-    .. automethod:: with_params
-    .. automethod:: get_predicted_times
+    * To evaluate the calibrated models, pass the kernel-specific calibration
+      parameters from :meth:`estimate_kernel_specific_calibration_params` to
+      :meth:`qbx_cost_per_stage` or :meth:`qbx_cost_per_box`.
     """
 
-    def __init__(self, raw_costs, params):
-        self.raw_costs = OrderedDict(raw_costs)
-        self.params = params
-
-    def with_params(self, params):
-        """Return a copy of *self* with parameters updated to include *params*."""
-        new_params = self.params.copy()
-        new_params.update(params)
-        return type(self)(
-                raw_costs=self.raw_costs.copy(),
-                params=new_params)
-
-    def copy(self):
-        return self.with_params({})
-
-    def __str__(self):
-        return "".join([
-                type(self).__name__,
-                "(raw_costs=",
-                str(self.raw_costs),
-                ", params=",
-                str(self.params),
-                ")"])
-
-    def __repr__(self):
-        return "".join([
-                type(self).__name__,
-                "(raw_costs=",
-                repr(self.raw_costs),
-                ", params=",
-                repr(self.params),
-                ")"])
-
-    def get_predicted_times(self, merge_close_lists=False):
-        """Return a dictionary mapping stage names to predicted time in seconds.
-
-        :arg merge_close_lists: If *True*, the returned estimate combines
-            the cost of "close" lists (Lists 1, 3 close, and 4 close). If
-            *False*, the time of each "close" list is reported separately.
+    @abstractmethod
+    def process_form_qbxl(self, queue, geo_data, p2qbxl_cost,
+                          ndirect_sources_per_target_box):
         """
-        from pymbolic import evaluate
-        from functools import partial
-
-        get_time = partial(evaluate, context=self.params)
-
-        result = OrderedDict()
-
-        for name, val in self.raw_costs.items():
-            if merge_close_lists:
-                for suffix in ("_list1", "_list3", "_list4"):
-                    if name.endswith(suffix):
-                        name = name[:-len(suffix)]
-                        break
-
-            result[name] = get_time(val) + result.get(name, 0)
-
-        return result
-
-# }}}
-
-
-# {{{ cost model
-
-class CostModel(object):
-    """
-    .. automethod:: with_calibration_params
-    .. automethod:: __call__
+        :arg queue: a :class:`pyopencl.CommandQueue` object.
+        :arg geo_data: a :class:`pytential.qbx.geometry.QBXFMMGeometryData` object or
+            similar object in the host memory.
```

Comment:
> in host memory

Is that intentional/true, even for GPU eval of the cost model?

Reply:
I intended to say the argument should be an object of class `QBXFMMGeometryData` for OpenCL cost model, but the same object in the host memory for Python cost model. But since we are positioning the Python as a test only class, I will just remove the "in host memory" part.